// slack-api-client/lib/api/groups.js
// Generated by CoffeeScript 1.10.0
(function() {
  var BaseClass, Groups,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  BaseClass = require('./common/base_class');

  Groups = (function(superClass) {
    extend(Groups, superClass);

    function Groups() {
      return Groups.__super__.constructor.apply(this, arguments);
    }

    Groups.prototype.endpoints = {
      archive: '/api/groups.archive',
      close: '/api/groups.close',
      create: '/api/groups.create',
      createChild: '/api/groups.createChild',
      history: '/api/groups.history',
      invite: '/api/groups.invite',
      kick: '/api/groups.kick',
      leave: '/api/groups.leave',
      list: '/api/groups.list',
      mark: '/api/groups.mark',
      open: '/api/groups.open',
      rename: '/api/groups.rename',
      setPurpose: '/api/groups.setPurpose',
      setTopic: '/api/groups.setTopic',
      unarchive: '/api/groups.unarchive'
    };

    Groups.prototype.archive = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.archive, opts, callback);
    };

    Groups.prototype.close = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.close, opts, callback);
    };

    Groups.prototype.create = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['name']);
      return this.request('POST', this.endpoints.create, opts, callback);
    };

    Groups.prototype.createChild = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.create, opts, callback);
    };

    Groups.prototype.history = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('GET', this.endpoints.history, opts, callback);
    };

    Groups.prototype.invite = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'user']);
      return this.request('POST', this.endpoints.invite, opts, callback);
    };

    Groups.prototype.kick = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'user']);
      return this.request('POST', this.endpoints.kick, opts, callback);
    };

    Groups.prototype.leave = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.leave, opts, callback);
    };

    Groups.prototype.list = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      return this.request('GET', this.endpoints.list, opts, callback);
    };

    Groups.prototype.mark = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'ts']);
      return this.request('POST', this.endpoints.mark, opts, callback);
    };

    Groups.prototype.open = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.open, opts, callback);
    };

    Groups.prototype.rename = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'name']);
      return this.request('POST', this.endpoints.rename, opts, callback);
    };

    Groups.prototype.setPurpose = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'purpose']);
      return this.request('POST', this.endpoints.setPurpose, opts, callback);
    };

    Groups.prototype.setTopic = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'topic']);
      return this.request('POST', this.endpoints.setTopic, opts, callback);
    };

    Groups.prototype.unarchive = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.unarchive, opts, callback);
    };

    return Groups;

  })(BaseClass);

  module.exports = Groups;

}).call(this);
