// slack-api-client/lib/api/mpim.js
// Generated by CoffeeScript 1.10.0
(function() {
  var BaseClass, MPIM,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  BaseClass = require('./common/base_class');

  MPIM = (function(superClass) {
    extend(MPIM, superClass);

    function MPIM() {
      return MPIM.__super__.constructor.apply(this, arguments);
    }

    MPIM.prototype.endpoints = {
      close: '/api/mpim.close',
      history: '/api/mpim.history',
      list: '/api/mpim.list',
      mark: '/api/mpim.mark',
      open: '/api/mpim.open'
    };

    MPIM.prototype.close = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('POST', this.endpoints.close, opts, callback);
    };

    MPIM.prototype.history = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel']);
      return this.request('GET', this.endpoints.history, opts, callback);
    };

    MPIM.prototype.list = function(callback) {
      return this.request('GET', this.endpoints.list, {}, callback);
    };

    MPIM.prototype.mark = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['channel', 'ts']);
      return this.request('POST', this.endpoints.mark, opts, callback);
    };

    MPIM.prototype.open = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['users']);
      return this.request('POST', this.endpoints.open, opts, callback);
    };

    return MPIM;

  })(BaseClass);

  module.exports = MPIM;

}).call(this);
