// slack-api-client/lib/api/search.js
// Generated by CoffeeScript 1.10.0
(function() {
  var BaseClass, Search,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  BaseClass = require('./common/base_class');

  Search = (function(superClass) {
    extend(Search, superClass);

    function Search() {
      return Search.__super__.constructor.apply(this, arguments);
    }

    Search.prototype.endpoints = {
      all: '/api/search.all',
      files: '/api/search.files',
      messages: '/api/search.messages'
    };

    Search.prototype.all = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['query']);
      return this.request('GET', this.endpoints.all, opts, callback);
    };

    Search.prototype.files = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['query']);
      return this.request('GET', this.endpoints.files, opts, callback);
    };

    Search.prototype.messages = function(opts, callback) {
      if (opts == null) {
        opts = {};
      }
      this.validate.obj(opts, ['query']);
      return this.request('POST', this.endpoints.messages, opts, callback);
    };

    return Search;

  })(BaseClass);

  module.exports = Search;

}).call(this);
